scipy/spatial/_ckdtree.pyx

```diff
@@ -791,7 +791,7 @@ cdef class cKDTree:
         if p < 1:
             raise ValueError("Only p-norms with 1<=p<=infinity permitted")
 
-        if not np.isfinite(x).all():
+        if not np.isfinite(x_arr).all():
```

Comment:
I'm confused, `query()` in `scipy/spatial/_kdtree.py` already has `x = np.asarray(x)`, so I'm not seeing how this will help? I put debug prints in to the original reproducer, and `x` is already coerced in the control flow on `main`, so what am I missing here?

In fact, the original reproducer passes when I test it locally with SciPy `main` and recent pandas:

```diff
--- a/scipy/spatial/tests/test_kdtree.py
+++ b/scipy/spatial/tests/test_kdtree.py
@@ -1488,3 +1488,16 @@ def test_nonfinite_inputs_gh_18223():
         KDTree(coords, balanced_tree=True, compact_nodes=True)
     with pytest.raises(ValueError, match="must be finite"):
         KDTree(coords, balanced_tree=False, compact_nodes=False)
+
+
+def test_gh_18800():
+    import pandas as pd
+
+    points = [
+        [66.22, 32.54],
+        [22.52, 22.39],
+        [31.01, 81.21],
+        ]
+
+    tree = KDTree(points, 10)
+    tree.query(pd.DataFrame(points), 1)
```

Reply:
Oh, this only happens when directly calling `cKDTree` -- isn't that a different kind of bug then? They're not supposed to have different behaviors anymore are they?

Reply:
Should the `x` be coerced in one control flow and not the other is what I'm asking I think? We can depend on it being coerced in the currently-encouraged `KDTree` call vs. discouraged `cKDTree` call.

Reply:
Or should we just refuse to fix bugs for `cKDTree` direct calls at this point? Our docs clearly indicate that `KDTree` direct calls should be preferred. This one is pretty easy, but maintaining both probably isn't something we want super long-term.

Reply:
I think we should fix bugs for cKDTree direct calls as long as it is exported in the scipy.spatial namespace. What we have here is a regression that is easy to fix. We could consider to deprecate (direct calls to) cKDTree though.